### ISSUE TITLE
Handle training sets that don't fit in memory

### ISSUE THREAD

**Issue by macfergus** [maxpumperla/betago#8]
I made two attempts at this:

1. Load a model from disk, load a small training set, run a few epochs, and write back to disk. Repeat. This actually worked kinda OK; after a couple hours it reached something like 0.15 training set accuracy, 0.20 validation set accuracy before it seems to top out. (Sorry, I wasn't very scientific and didn't track the exact number of training examples and epochs.)

2. Use `fit_generator` to train on a chunked data set. I ran a few epochs on 10000 samples (which yields about 1900000 training examples) and the results weren't very exciting; I'm trying to run lots more epochs on a larger set to see what happens.

I pushed some REALLY rough code to a branch in case someone wants to pick it up. (I probably won't have to time to continue on it for a few days)

https://github.com/macfergus/betago/blob/memory-hacks/examples/update.py
https://github.com/macfergus/betago/blob/memory-hacks/betago/dataloader/base_processor.py#L30

**Comment by maxpumperla** [comment]
Nice, I wouldn't care too much about model performance right now. Probably we also have to care about the network architecture itself more. I really like 2. above. If you don't mind, I would pick this up tomorrow or so. 

This is the way to go for now, but there's a few other things that could be interesting: 
a) Load data using dask and 
b) parallelize everything with Spark using elephas. Putting data on S3 and generating an RDD is straight forward.

**Comment by macfergus** [comment]
Yes please feel free to continue on it! Apologies for the super-sloppy code. A couple notes:

- I don't think all the calls to gc.collect() are actually necessary (I put them in while debugging something)
- It breaks up the biggest KGS files into chunks, see here:
https://github.com/macfergus/betago/blob/memory-hacks/betago/dataloader/base_processor.py#L332
512 examples per chunk is probably smaller than necessary.
- It doesn't clean up after itself properly, you may have to clear all the .npy files out of your data dir between runs

**Comment by maxpumperla** [comment]
Absolutely no need to apologize. Apart from the fact that it's not even sloppy in my eyes, we're here to build fun stuff in our free time - no place for apologies! :) Your effort is much appreciated, thanks again!

**Comment by maxpumperla** [comment]
Hey, @macfergus, I've integrated your ideas in master. It's not cleaned-up yet, but works for now. I've opted for a ```use_generator()``` flag in ```GoDataProcessor```, so that both numpy arrays and generators can be returned. This way we don't need  ```generate_go_data``` as new method, and the parts collecting data stay contained in ```consolidate_games```, as initially intended.

I pushed it so that you could have look whenever you want. 

https://github.com/maxpumperla/betago/blob/master/examples/train_generator.py

**Comment by macfergus** [comment]
Awesome! Let's close this issue. There's still lots of ways to improve the data loading, but this should unblock anyone who just wants to play with a bigger sample.
